src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -374,31 +374,20 @@ private void CheckSnapshot(ITextSnapshot snapshot)
 
             private ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> ConvertToTagTrees(
                 ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>> oldTagTrees,
+                ISet<ITextBuffer> buffersToTag,
                 ILookup<ITextBuffer, ITagSpan<TTag>> newTagsByBuffer,
                 IEnumerable<DocumentSnapshotSpan> spansTagged)
             {
-                // NOTE: we assume that the following list is already realized and is _not_ lazily
-                // computed. It's not clear what the contract is of this API.
-
-                // common case where we only tagged a single range of a document.
-                if (spansTagged.IsSingle())
-                {
-                    return ConvertToTagTree(oldTagTrees, newTagsByBuffer, spansTagged.Single().SnapshotSpan);
-                }
-
-                // heavy linq case 
-                var spansToInvalidateByBuffer = spansTagged.Select(ss => ss.SnapshotSpan).ToLookup(ss => ss.Snapshot.TextBuffer);
-
-                var buffers = oldTagTrees.Keys.Concat(newTagsByBuffer.Select(g => g.Key))
-                                              .Concat(spansTagged.Select(dss => dss.SnapshotSpan.Snapshot.TextBuffer))
-                                              .Distinct();
+                var spansToInvalidateByBuffer = spansTagged.ToLookup(
+                    keySelector: span => span.SnapshotSpan.Snapshot.TextBuffer,
+                    elementSelector: span => span.SnapshotSpan);
 
                 // Walk through each relevant buffer and decide what the interval tree should be
                 // for that buffer.  In general this will work by keeping around old tags that
                 // weren't in the range that was re-tagged, and merging them with the new tags
                 // produced for the range that was re-tagged.
                 var newTagTrees = ImmutableDictionary<ITextBuffer, TagSpanIntervalTree<TTag>>.Empty;
-                foreach (var buffer in buffers)
+                foreach (var buffer in buffersToTag)
                 {
                     var newTagTree = ComputeNewTagTree(oldTagTrees, buffer, newTagsByBuffer[buffer], spansToInvalidateByBuffer[buffer]);
```

Comment:
so we basically recalculate all tags every time and never tries to reuse some? is that okay with the view port only tagger thingy?